### PR TITLE
fix apiserver start failed if lost one of cert and key

### PR DESCRIPTION
**Pull request by xilabao**
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36472)

### CODE REVIEW

**Comment by dims** [comment]
@k8s-bot ok to test

**Comment by liggitt** [comment]
I don't think we want to half-bootstrap… if only one of the files exists, that probably means something went wrong. A better error message might be good, but I wouldn't want to assume we can overwrite the one existing file with a self-generated one

**Comment by xilabao** [comment]
@liggitt PTAL

**Comment by liggitt** [comment]
what is the error message without this change?

**Comment by xilabao** [comment]
the original error message is "E1117 11:30:50.075136   26608 server.go:254] Failed to create clientset: open /var/run/kubernetes/apiserver.crt: no such file or directory", Maybe we can stop apiserver at the checkpoint and output a more clear message.

**Comment by lavalamp** [comment]
/lgtm

@liggitt I agree with you but this appears to error out instead of generating anything in that case. Am I reading it wrong?

**Comment by liggitt** [comment]
No, it got reworked after my comment

File: pkg/genericapiserver/config.go [hunk 1]
```diff
@@ -406,7 +406,12 @@ func (c completedConfig) New() (*GenericAPIServer, error) {
 func (c completedConfig) MaybeGenerateServingCerts(alternateIPs ...net.IP) error {
 	// It would be nice to set a fqdn subject alt name, but only the kubelets know, the apiserver is clueless
 	// alternateDNS = append(alternateDNS, "kubernetes.default.svc.CLUSTER.DNS.NAME")
-	if c.SecureServingInfo != nil && c.SecureServingInfo.ServerCert.Generate && !certutil.CanReadCertOrKey(c.SecureServingInfo.ServerCert.CertFile, c.SecureServingInfo.ServerCert.KeyFile) {
+	canReadCertAndKey, err := certutil.CanReadCertAndKey(c.SecureServingInfo.ServerCert.CertFile, c.SecureServingInfo.ServerCert.KeyFile)
```
Comment (sttts) [inline pkg/genericapiserver/config.go]: The logic is wrong: if secure serving is disabled (`c.SecureServingInfo == nil`) we still check for cert and key.

**Comment by sttts** [comment]
Please fix the logic and remove the panic on `c.SecureServingInfo == nil`.

File: pkg/util/cert/io.go [hunk 2]
```diff
@@ -25,14 +25,25 @@ import (
...
-		return true
+func CanReadCertAndKey(certPath, keyPath string) (bool, error) {
+	certExist := canReadFile(certPath)
+	keyExist := canReadFile(keyPath)
+
+	if certExist == false && keyExist == false {
+		return false, nil
+	}
+
+	if certExist == false {
```
Comment (xilabao) [inline pkg/util/cert/io.go]: The logic here is match your mean. In the end, will return true.

File: pkg/genericapiserver/config.go (hunk 1 above)
Comment (xilabao) [inline pkg/genericapiserver/config.go]: I will fix it. Thanks

File: pkg/util/cert/io.go (hunk 2 above)
Comment (sttts) [inline pkg/util/cert/io.go]: By the godoc and the function name, it should return false if one of the certs cannot be read. It returns an error instead.

Comment (sttts) [inline pkg/util/cert/io.go]: Maybe we can even inline this code. There is no added value of the whole function IMO.

**Comment by xilabao** [comment]
@sttts PTAL thanks

File: pkg/genericapiserver/config.go [hunk 3]
```diff
@@ -406,21 +406,27 @@ func (c completedConfig) New() (*GenericAPIServer, error) {
...
 	// alternateDNS = append(alternateDNS, "kubernetes.default.svc.CLUSTER.DNS.NAME")
-	if c.SecureServingInfo != nil && c.SecureServingInfo.ServerCert.Generate && !certutil.CanReadCertOrKey(c.SecureServingInfo.ServerCert.CertFile, c.SecureServingInfo.ServerCert.KeyFile) {
-		// TODO (cjcullen): Is ClusterIP the right address to sign a cert with?
-		alternateDNS := []string{"kubernetes.default.svc", "kubernetes.default", "kubernetes", "localhost"}
+	if c.SecureServingInfo != nil && c.SecureServingInfo.ServerCert.Generate {
+		canReadCertAndKey, err := certutil.CanReadCertAndKey(c.SecureServingInfo.ServerCert.CertFile, c.SecureServingInfo.ServerCert.KeyFile)
+		if err != nil {
+			return err
+		}
+		if !canReadCertAndKey {
```
Comment (sttts) [inline pkg/genericapiserver/config.go]: if canReadCertAndKey { return nil }

Comment (xilabao) [inline pkg/genericapiserver/config.go]: Done

File: pkg/util/cert/io.go [hunk 4]
```diff
@@ -25,14 +25,25 @@ import (
...
 
-// CanReadCertOrKey returns true if the certificate or key files already exists,
-// otherwise returns false.
-func CanReadCertOrKey(certPath, keyPath string) bool {
-	if canReadFile(certPath) || canReadFile(keyPath) {
-		return true
+// CanReadCertAndKey returns true if the certificate and key files already exists,
+// otherwise returns false. If lost one of cert and key, returns error.
+func CanReadCertAndKey(certPath, keyPath string) (bool, error) {
+	certExist := canReadFile(certPath)
```
Comment (sttts) [inline pkg/util/cert/io.go]: s/Exist/Readable/

Comment (xilabao) [inline pkg/util/cert/io.go]: Done

**Comment by sttts** [comment]
Lgtm, thanks for fixing.

**Pull request merged**